src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -182,21 +185,45 @@ public bool HandlePartialDiscoveryComplete(IProxyDiscoveryManager proxyDiscovery
 
     private List<ProviderSpecificWorkload<DiscoveryCriteria>> SplitToWorkloads(DiscoveryCriteria discoveryCriteria, Dictionary<string, TestRuntimeProviderInfo> sourceToTestHostProviderMap)
     {
+        var sources = discoveryCriteria.Sources;
+        // Each source is grouped with its respective provider.
+        var providerGroups = sources
+            .Select(source => new ProviderSpecificWorkload<string>(source, sourceToTestHostProviderMap[source]))
+            .GroupBy(psw => psw.Provider);
+
         List<ProviderSpecificWorkload<DiscoveryCriteria>> workloads = new();
-        foreach (var source in discoveryCriteria.Sources)
+        foreach (var group in providerGroups)
         {
-            var testHostProviderInfo = sourceToTestHostProviderMap[source];
-            var runsettingsXml = testHostProviderInfo.RunSettings;
-            var updatedDiscoveryCriteria = new ProviderSpecificWorkload<DiscoveryCriteria>(NewDiscoveryCriteriaFromSourceAndSettings(source, discoveryCriteria, runsettingsXml), testHostProviderInfo);
-            workloads.Add(updatedDiscoveryCriteria);
+            var testhostProviderInfo = group.Key;
+            // If the run is not parallel and the host is shared, put all testcases on single testhost.
+            if (!_isParallel && testhostProviderInfo.Shared)
+            {
+                var runsettings = testhostProviderInfo.RunSettings;
+                var sourcesToDiscover = group.Select(w => w.Work).ToArray();
+                var updatedCriteria = NewDiscoveryCriteriaFromSourceAndSettings(sourcesToDiscover, discoveryCriteria, runsettings);
+                var workload = new ProviderSpecificWorkload<DiscoveryCriteria>(updatedCriteria, testhostProviderInfo);
+                workloads.Add(workload);
+            }
+            else
+            {
+                // Create one workload for each source
+                foreach (var w in group.ToList())
+                {
+                    var runsettings = testhostProviderInfo.RunSettings;
+                    var sourcesToDiscover = new[] { w.Work };
+                    var updatedCriteria = NewDiscoveryCriteriaFromSourceAndSettings(sourcesToDiscover, discoveryCriteria, runsettings);
+                    var workload = new ProviderSpecificWorkload<DiscoveryCriteria>(updatedCriteria, testhostProviderInfo);
+                    workloads.Add(workload);
+                }
+            }
         }
 
         return workloads;
 
-        static DiscoveryCriteria NewDiscoveryCriteriaFromSourceAndSettings(string source, DiscoveryCriteria discoveryCriteria, string runsettingsXml)
+        static DiscoveryCriteria NewDiscoveryCriteriaFromSourceAndSettings(string[] sources, DiscoveryCriteria discoveryCriteria, string runsettingsXml)
```

Comment:
This could be `IEnumerable<string>` instead of array (this is the "level" expected by `DiscoveryCriteria`.

Reply:
Or teach me :)